design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,443 @@
+both by a user when changing the settings or as a part of input validation.
+The BIOS Registry support was recently added to Ironic (see the Ironic proposal
+[BIOS Registry](https://review.opendev.org/c/openstack/ironic-specs/+/774681).
+This schema is the same for each vendor's model so it will not need to be
+updated when stored. More info for the BIOS registry data can be found in the
+[Redfish Registry](https://redfish.dmtf.org/schemas/v1/AttributeRegistry.v1_3_5.json)
+
+The schema fields returned via the BIOS API include:
+
+- Type (e.g. enum, string, integer, boolean, or password)
+- ReadOnly (a boolean indicating whether the attribute is read only)
+- Unique (indicating the value is specific to the host where it was
+  retrieved)
+- Allowable values for enum type attributes
+- Minimum and maximum values (for numeric attributes)
+- Minimum and maximum character lengths (for string attributes)
+
+The BIOS Settings are retrieved from the BMC by Ironic and cached
+whenever the node moves to `manageable` or `cleaning`, or when the settings
+are updated. The baremetal-operator will manage the data as follows:
+
+- The node first transitions to manageable during the `Registering` state, so
+  at the end of that state `settingsData` will be populated.
+- Settings can be updated during the `Preparing` state, so at the end of that
+  state the settings will also be retrieved and used to update
+  `settingsData`.
+
+Upon retrieving the BIOS Settings from Ironic, the baremetal-operator will
+copy the name/value pairs to the `vendorSettings` field in
+`HostFirwareSettingsSpec` if that field is empty. Settings that have the
+`read-only` flag set will not be copied as these settings cannot be written
+to the BMC. Likewise, settings that have the `unique` flag set will also not
+be copied, these are settings like serial numbers or product ids for example,
+that should not be written to other servers.
+
+A user can update `vendorSettings` to the desired values for the
+settings using the limits from the `FirmwareSchema` section as a guide.
+
+The baremetal-operator will detect when `vendorSettings` changes by comparing
+the name/value pairs to `settingsData`. When a change is detected, the BMO
+will add the new values to the Ironic clean-steps API in the Preparing state,
+when building the manual clean steps. Before adding it though, the BMO will
+do validation checks on the new values using the schema. For example, integer
+types will checked against minimum and maximum parameters and enumeration types
+will be compared against allowable values. A failure will be returned from the
+clean steps stage if any errors are detected in order to avoid pushing a
+partial configuration to Ironic.
+
+As a follow-on, a validation webhook for the HostFirmwareSettings CRD can be
+leveraged to add the validations. This webhook would reject the update before
+it is stored in the resource, removing the need to do the validations in
+manual clean steps.
```

Comment:
As per webhook best practices (see ["Validating a resource in the controller"](https://www.openshift.com/blog/kubernetes-operators-best-practices)), there could be some scenario where the webhook could not be available, so it'd be recommended to apply the same validation rules still in the controller itself (using some shared code)

Reply:
Updated text to indicate that the same checks will be done before adding clean steps.